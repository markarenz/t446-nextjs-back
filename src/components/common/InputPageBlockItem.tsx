import { useState, useMemo, useEffect } from 'react';
import { getBlockDefByType } from '../../helpers/pages';
import { Asset } from '../../types/types';
import IconButton from './IconButton';
import { IconClose, IconUp, IconDown, IconInfo } from '../img/icons';
import Input from './Input';
import InputMarkdown from './InputMarkdown';
import InputDropdown from './InputDropdown';
import InputImagePicker from './InputImagePicker';
import { debounce } from 'lodash';
import { sanitizeString } from '../../helpers';
import styles from '../../styles/modules/InputPageBlockItem.module.scss';

type Props = {
  idx: number;
  block: any;
  isLastBlock: boolean;
  moveBlock: Function;
  removeBlock: Function;
  updateBlocks: Function;
  baseImgUrl: string;
  assetList: Asset[];
  handleTriggerRefresh: Function;
};
const InputPageBlockItem: React.FC<Props> = ({
  idx,
  block,
  moveBlock,
  removeBlock,
  isLastBlock,
  updateBlocks,
  baseImgUrl,
  assetList,
  handleTriggerRefresh
}) => {
  const [showIdClass, setShowIdClass] = useState(false);
  const [localData, setLocalData] = useState(block);
  const blockDef = getBlockDefByType(block.type);
  const moveUp = () => {
    moveBlock(idx, -1);
  };
  const moveDown = () => {
    moveBlock(idx, 1);
  };
  const removeThisBlock = () => {
    removeBlock(idx);
  };
  const callUpdateBlocks = () => {
    updateBlocks(idx, localData);
  };
  const debouncedUpdateBlocks = useMemo(
    () => debounce(callUpdateBlocks, 300),
    // eslint-disable-next-line react-hooks/exhaustive-deps
    [localData]
  );
  const handleToggleShowIdClass = () => {
    setShowIdClass((prevShowIdClass) => !prevShowIdClass);
  };
  const handleChange = (e: React.ChangeEvent<HTMLInputElement>) => {
    setLocalData((prevLocalData: any) => ({
      ...prevLocalData,
      [e.target.name]: sanitizeString(e.target.value)
    }));
  };
  const handleChangeValue = (name: string, value: string) => {
    setLocalData((prevLocalData: any) => ({
      ...prevLocalData,
      [name]: sanitizeString(value)
    }));
  };
  useEffect(() => {
    debouncedUpdateBlocks();
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [localData]);
  return (
    <div className={styles.root} data-testid={`input-page-block-item-${idx}`}>
      <h3 className={styles.title}>
        {blockDef?.title || ''}
        <div>
          <span className="mr-1">
            <IconButton
              title="More detail"
              disabled={false}
              onClick={handleToggleShowIdClass}
              color="secondary"
              testId="block-btn-info"
            >
              <IconInfo />
            </IconButton>
          </span>
          <span className="mr-1">
            <IconButton
              title="Move Up"
              disabled={idx < 1}
              onClick={moveUp}
              color="default"
              testId="block-btn-move-up"
            >
              <IconUp />
            </IconButton>
          </span>
          <span className="mr-1">
            <IconButton
              title="Move Down"
              disabled={isLastBlock}
              onClick={moveDown}
              color="default"
              testId="block-btn-move-down"
            >
              <IconDown />
            </IconButton>
          </span>
          <IconButton
            title="Remove"
            onClick={removeThisBlock}
            color="primary"
            testId="block-btn-remove"
          >
            <IconClose />
          </IconButton>
        </div>
      </h3>
      <div className="grid">
        {showIdClass && (
          <div className="col-6" data-testid="page-block-info">
            <label>ID</label>
            <Input
              type="text"
              name="id"
              onChange={handleChange}
              autoFocus={false}
            />
          </div>
        )}
        {showIdClass && (
          <div className="col-6">
            <label>Class</label>
            <Input
              type="text"
              name="class"
              onChange={handleChange}
              autoFocus={false}
            />
          </div>
        )}
        {blockDef?.fields.map((f, idx) => (
          <div key={`${f.slug}-${idx}`} className={`col-${f.colsize}`}>
            <label>{f.label}</label>
            {['markdownarea', 'markdown'].includes(f.type) && (
              <InputMarkdown
                name={f.slug}
                value={localData[f.slug]}
                onChange={handleChange}
                rows={5}
                tabIndex={0}
                testId={`block-${idx}-${f.slug}`}
              />
            )}
            {f.type === 'text' && (
              <div>
                <Input
                  name={f.slug}
                  autoFocus={false}
                  value={localData[f.slug]}
                  onChange={handleChange}
                  testId={`block-${idx}-${f.slug}`}
                />
              </div>
            )}
            {f.type === 'select' && (
              <InputDropdown
                name={f.slug}
                value={localData[f.slug]}
                onChange={handleChange}
                options={f?.options || []}
                testId={`block-${idx}-${f.slug}`}
              />
            )}
            {f.type === 'image' && (
              <InputImagePicker
                name={f.slug}
                value={localData[f.slug]}
                onChange={handleChangeValue}
                baseImgUrl={`${baseImgUrl}`}
                assetList={assetList}
                handleTriggerRefresh={handleTriggerRefresh}
              />
            )}
          </div>
        ))}
      </div>
    </div>
  );
};

export default InputPageBlockItem;
