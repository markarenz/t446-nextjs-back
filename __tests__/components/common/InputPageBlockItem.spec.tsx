import React from 'react';
import { render, screen, fireEvent, waitFor } from '@testing-library/react';
import { getAssetList } from '../../__fixtures__/assetList';
import InputPageBlockItem from '../../../src/components/common/InputPageBlockItem';

jest.mock('react-markdown', () => () => <div>This is my __markdown__.</div>);
jest.mock('lodash', () => ({
  debounce: jest.fn((cb) => cb)
}));

beforeEach(() => {
  jest.clearAllMocks();
});

const assetList = getAssetList();

const block1 = {
  type: 'cta-one',
  button_link: '/contact',
  button_text: 'Contact',
  class: '',
  headline: 'Adventure Awaits',
  id: '',
  text: 'Text goes here'
};
const block2 = {
  type: '2-col-photo',
  align: 'left',
  class: '',
  html: 'test text',
  id: '',
  photo: 't446-caving-01.jpg',
  tcol: 'right'
};
const defaultProps = {
  idx: 1,
  block: block1,
  isLastBlock: false,
  moveBlock: jest.fn(),
  removeBlock: jest.fn(),
  updateBlocks: jest.fn(),
  baseImgUrl: 'https://www.domain.dom/',
  assetList,
  handleTriggerRefresh: jest.fn()
};
const altProps = {
  ...defaultProps,
  block: block2
};
describe('InputPageBlockItem', () => {
  it('renders default component', async () => {
    render(<InputPageBlockItem {...defaultProps} />);
    const element = await screen.findByTestId('input-page-block-item-1');
    expect(element).toBeInTheDocument();
    expect(element).toMatchSnapshot();
  });
  it('toggles info', async () => {
    render(<InputPageBlockItem {...defaultProps} />);
    const info = screen.queryByTestId('page-block-info');
    const element = await screen.findByTestId('block-btn-info');
    expect(element).toBeInTheDocument();
    expect(info).not.toBeInTheDocument();
    await waitFor(() =>
      fireEvent(
        element,
        new MouseEvent('click', {
          bubbles: true,
          cancelable: true
        })
      )
    );
    const info2 = screen.queryByTestId('page-block-info');
    expect(info2).toBeInTheDocument();
  });

  it('moves block up', async () => {
    render(<InputPageBlockItem {...defaultProps} />);
    const btnUp = await screen.findByTestId('block-btn-move-up');
    await waitFor(() =>
      fireEvent(
        btnUp,
        new MouseEvent('click', {
          bubbles: true,
          cancelable: true
        })
      )
    );
    expect(defaultProps.moveBlock).toHaveBeenCalled();
  });

  it('moves block down', async () => {
    render(<InputPageBlockItem {...defaultProps} />);
    const btnDown = await screen.findByTestId('block-btn-move-down');
    await waitFor(() =>
      fireEvent(
        btnDown,
        new MouseEvent('click', {
          bubbles: true,
          cancelable: true
        })
      )
    );
    expect(defaultProps.moveBlock).toHaveBeenCalled();
  });

  it('handles input updates', async () => {
    render(<InputPageBlockItem {...defaultProps} />);
    const input = await screen.findByTestId('block-0-headline');
    await waitFor(() =>
      fireEvent.change(input, {
        target: { value: 'New text' }
      })
    );
    expect(defaultProps.updateBlocks).toHaveBeenCalled();
  });
  it('handles value updates', async () => {
    render(<InputPageBlockItem {...altProps} />);
    //
    const input = await screen.findByTestId('input-image-picker-opener');
    await waitFor(() =>
      fireEvent(
        input,
        new MouseEvent('click', {
          bubbles: true,
          cancelable: true
        })
      )
    );
    const assetItem = await screen.findByTestId('asset-item-0');
    await waitFor(() =>
      fireEvent(
        assetItem,
        new MouseEvent('click', {
          bubbles: true,
          cancelable: true
        })
      )
    );
    expect(defaultProps.updateBlocks).toHaveBeenCalled();
  });

  it('calls remove function when remove button clicked', async () => {
    render(<InputPageBlockItem {...defaultProps} />);
    const element = await screen.findByTestId('block-btn-remove');
    await waitFor(() =>
      fireEvent(
        element,
        new MouseEvent('click', {
          bubbles: true,
          cancelable: true
        })
      )
    );
    expect(defaultProps.removeBlock).toHaveBeenCalled();
  });
});
