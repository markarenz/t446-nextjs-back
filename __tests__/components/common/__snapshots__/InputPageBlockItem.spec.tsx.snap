// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`InputPageBlockItem renders default component 1`] = `
<div
  class="root"
  data-testid="input-page-block-item-1"
>
  <h3
    class="title"
  >
    Call to action
    <div>
      <span
        class="mr-1"
      >
        <button
          class="iconButton secondary"
          data-testid="block-btn-info"
          title="More detail"
          type="button"
        >
          <div
            class="inner"
          >
            <svg
              version="1.1"
              viewBox="0 0 100 100"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M36.228 79.507q0-1.292.62-4.082.362-1.602 1.086-4.238l6.356-23.77q.103-.414.155-.828.103-.413.103-.775 0-2.119-1.343-2.635-1.344-.517-4.91-.672v-2.79q2.998-.207 8.682-.827 5.736-.672 8.113-1.137l4.548-.879-8.217 30.127q-1.033 3.876-1.395 5.375-.93 3.824-.93 4.96 0 1.138.516 1.5.517.31 1.137.31 1.602 0 3.98-2.74 2.376-2.738 4.185-5.735l2.326 1.498q-5.168 7.597-7.907 10.232-4.496 4.34-9.612 4.34-2.894 0-5.219-1.756-2.274-1.809-2.274-5.478zm20.464-66.249q2.946 0 5.013 2.067 2.067 2.067 2.067 5.013 0 2.945-2.067 5.064-2.067 2.067-5.013 2.067-2.945 0-5.064-2.067-2.067-2.119-2.067-5.064 0-2.946 2.067-5.013 2.119-2.067 5.064-2.067z"
                font-family="Times"
                font-style="italic"
                font-weight="bold"
              />
            </svg>
          </div>
        </button>
      </span>
      <span
        class="mr-1"
      >
        <button
          class="iconButton default"
          data-testid="block-btn-move-up"
          title="Move Up"
          type="button"
        >
          <div
            class="inner"
          >
            <svg
              version="1.1"
              viewBox="0 0 100 100"
              xmlns="http://www.w3.org/2000/svg"
            >
              <g>
                <path
                  d="M87.5 77.5L95 70 50 25 5 70l7.5 7.5L50 40z"
                  paint-order="markers fill stroke"
                  stroke-linecap="round"
                  stroke-width="0"
                />
              </g>
            </svg>
          </div>
        </button>
      </span>
      <span
        class="mr-1"
      >
        <button
          class="iconButton default"
          data-testid="block-btn-move-down"
          title="Move Down"
          type="button"
        >
          <div
            class="inner"
          >
            <svg
              version="1.1"
              viewBox="0 0 100 100"
              xmlns="http://www.w3.org/2000/svg"
            >
              <g>
                <path
                  d="M87.5 25l7.5 7.5-45 45-45-45 7.5-7.5L50 62.5z"
                  paint-order="markers fill stroke"
                  stroke-linecap="round"
                  stroke-width="0"
                />
              </g>
            </svg>
          </div>
        </button>
      </span>
      <button
        class="iconButton primary"
        data-testid="block-btn-remove"
        title="Remove"
        type="button"
      >
        <div
          class="inner"
        >
          <svg
            version="1.1"
            viewBox="0 0 100 100"
            xmlns="http://www.w3.org/2000/svg"
          >
            <g>
              <path
                d="M13.182 5L5 13.182 41.818 50 5 86.818 13.182 95 50 58.18 86.818 95 95 86.818 58.18 50 95 13.182 86.818 5 50 41.818z"
                fill="navy"
                paint-order="markers fill stroke"
                stroke="#fff"
                stroke-linecap="round"
                stroke-width="0"
              />
            </g>
          </svg>
        </div>
      </button>
    </div>
  </h3>
  <div
    class="grid"
  >
    <div
      class="col-12"
    >
      <label>
        Headline
      </label>
      <div>
        <input
          class="input"
          data-testid="block-0-headline"
          name="headline"
          value="Adventure Awaits"
        />
      </div>
    </div>
    <div
      class="col-12"
    >
      <label>
        Sub-Headline
      </label>
      <div
        class="root"
        data-testid="block-1-text"
      >
        <div
          class="modeToggleWrapper"
        >
          <button
            class="iconButton default"
            data-testid="input-markdown-toggle-preview"
            title="toggle preview"
            type="button"
          >
            <div
              class="inner"
            >
              <svg
                version="1.1"
                viewBox="0 0 80.794 80.794"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M79.351 38.549c-.706-.903-17.529-22.119-38.953-22.119-21.426 0-38.249 21.216-38.955 22.119L0 40.396l1.443 1.847c.706.903 17.529 22.12 38.955 22.12 21.424 0 38.247-21.217 38.953-22.12l1.443-1.847-1.443-1.847zM40.398 58.364c-15.068 0-28.22-13.046-32.643-17.967 4.425-4.922 17.576-17.966 32.643-17.966 15.066 0 28.218 13.045 32.642 17.966-4.426 4.922-17.577 17.967-32.642 17.967z"
                />
                <path
                  d="M40.397 23.983c-9.052 0-16.416 7.363-16.416 16.414 0 9.053 7.364 16.417 16.416 16.417s16.416-7.364 16.416-16.417c0-9.051-7.364-16.414-16.416-16.414zm0 26.83c-5.744 0-10.416-4.673-10.416-10.417 0-5.742 4.672-10.414 10.416-10.414 5.743 0 10.416 4.672 10.416 10.414 0 5.744-4.673 10.417-10.416 10.417z"
                />
              </svg>
            </div>
          </button>
        </div>
        <div
          class="grid"
        >
          <div
            class="col-12"
          >
            <textarea
              class="textarea"
              name="text"
              rows="5"
              tabindex="0"
            >
              Text goes here
            </textarea>
          </div>
        </div>
      </div>
    </div>
    <div
      class="col-6"
    >
      <label>
        Button Text
      </label>
      <div>
        <input
          class="input"
          data-testid="block-2-button_text"
          name="button_text"
          value="Contact"
        />
      </div>
    </div>
    <div
      class="col-6"
    >
      <label>
        Button Link
      </label>
      <div>
        <input
          class="input"
          data-testid="block-3-button_link"
          name="button_link"
          value="/contact"
        />
      </div>
    </div>
  </div>
</div>
`;
